docs/code/MlNetCookBook.md

```diff
@@ -0,0 +1,212 @@
+	    // We read the first 10 values as a single float vector.
+        new TextLoader.Column("FeatureVector", DataKind.R4, new[] {new TextLoader.Range(0, 9)}),
+        // Separately, read the target variable.
+        new TextLoader.Column("Target", DataKind.R4, 10)
+    },
+    // Default separator is tab, but we need a comma.
+    Separator = ","
+});
+
+// Now read the file. 
+var data = reader.Read(new MultiFileSource(dataPath));
+```
+
+## How do I look at the intermediate data?
+
+Oftentimes, when we construct the experiment, we want to make sure that the data processing 'up to a certain moment' produces the results that we want. With ML.NET it is not very easy to do: since all ML.NET operations are lazy, the objects we construct are just 'promises' of data.
+
+We will need to create the cursor and scan the data to obtain the actual values. One way to do this is to use [schema comprehension](SchemaComprehension.md) and map the data to an `IEnumerable` of user-defined objects.
+
+Example file (https://github.com/dotnet/machinelearning/blob/master/test/data/adult.tiny.with-schema.txt):
+```
+Label	Workclass	education	marital-status
+0	Private	11th	Never-married
+0	Private	HS-grad	Married-civ-spouse
+1	Local-gov	Assoc-acdm	Married-civ-spouse
+1	Private	Some-college	Married-civ-spouse
+
+```
+
+```c#
+// Create a new environment for ML.NET operations. It can be used for exception tracking and logging, 
+// as well as the source of randomness.
+var env = new LocalEnvironment();
+
+// Create the reader: define the data columns and where to find them in the text file.
+var reader = TextLoader.CreateReader(env, ctx => (
+        // A boolean column depicting the 'target label'.
+        IsOver50K: ctx.LoadBool(0),
+        // Three text columns.
+        Workclass: ctx.LoadText(1),
+        Education: ctx.LoadText(2),
+        MaritalStatus: ctx.LoadText(3)),
+    hasHeader: true);
+
+// Start creating our processing pipeline. For now, let's just concatenate all the text columns
+// together into one.
+var dataPipeline = reader.MakeNewEstimator()
+    .Append(row =>
+        (
+            row.IsOver50K,
+            AllFeatures: row.Workclass.ConcatWith(row.Education, row.MaritalStatus)
+        ));
+
+// Let's verify that the data has been read correctly. 
+// First, we read the data file.
+var data = reader.Read(new MultiFileSource(dataPath));
+
+// Fit our data pipeline and transform data with it.
+var transformedData = dataPipeline.Fit(data).Transform(data);
+
+// 'transformedData' is a 'promise' of data. Let's actually read it.
+var someRows = transformedData.AsDynamic
+    // Convert to an enumerable of user-defined type. 
+    .AsEnumerable<InspectedRow>(env, reuseRowObject: false)
+    // Take a couple values as an array.
+    .Take(4).ToArray();
+
+// Now we can inspect 'someRows' to see if the data has been read and transformed correctly.
+```
+
+The above code assumes that we defined our `InspectedRow` class as follows:
```

Comment:
In the above example `.Append(...)`, wouldn't all columns except `IsOver50K` and `AllFeatures` be dropped?

Reply:
No

---
In reply to: [219396248](https://github.com/dotnet/machinelearning/pull/960#discussion_r219396248) [](ancestors = 219396248)